Translate useInsertionEffect (#642)
reactjs/es.react.dev: beta/src/content/reference/react/useInsertionEffect.md

<Pitfall>

> `useInsertionEffect` está orientado a autores de librerías CSS-in-JS. A menos que estés trabajando en una librería CSS-in-JS y necesites un lugar donde inyectar los estilos, probablemente busques [`useEffect`](/reference/react/useEffect) o [`useLayoutEffect`](/reference/react/useLayoutEffect) en su lugar.

Reviewer: - En esta traducción solemos usar «biblioteca» como traducción de *«library»*.
- Creo que no habría daño en traducir CSS-in-JS a CSS-en-JS.
(Esto aplica a otras partes del documento en que se repiten estas situaciones).

---

## Referencia {/*reference*/}

### `useInsertionEffect(setup, dependencies?)` {/*useinsertioneffect*/}

Llama a `useInsertionEffect` para insertar los estilos antes de cualquier mutación en el DOM:

```js
import { useInsertionEffect } from 'react';

// En tu librería CSS-in-JS
function useCSS(rule) {
  useInsertionEffect(() => {
    // ... inyecta las etiquetas <style> aquí ...
  });
  return rule;
}
```

[Ver más ejemplos a continuación.](#usage)

#### Parámetros {/*parameters*/}

> * `setup`: La función con la lógica de tu Effect. Tu función setup puede opcionalmente devolver una función de *limpieza*. Antes de que tu componente sea añadido primero al DOM, React ejecutará tu función setup. Después de cada re-renderizado con dependencias modificadas, React ejecutará primero la función de limpieza (si es que la habías incluido) con los valores antiguos y entonces ejecutará tu función setup con los nuevos valores. Antes de que tu componente sea eliminado del DOM, React ejecutará tu función de limpieza una última vez.

Reviewer: Effect lo hemos traducido como Efecto en las demás páginas, respetando el uso de mayúsculas del original.

```suggestion
* `setup`: La función con la lógica de tu Effect. Tu función setup puede opcionalmente devolver una función de *limpieza*. Antes de que tu componente sea añadido primero al DOM, React ejecutará tu función setup. Después de cada re-renderizado con dependencias modificadas, React ejecutará primero la función de limpieza (si es que la habías incluido) con los valores antiguos y entonces ejecutará tu función setup con los nuevos valores. Antes de que tu componente sea eliminado del DOM, React ejecutará tu función de limpieza una última vez.
```

* **opcional** `dependencias`: La lista de todos los valores reactivos referenciados dentro del el código de `setup`. Los valores reactivos incluyen props, estado y todas las variables y funciones declaradas directamente dentro del cuerpo de tu componente. Si tu linter está [configurado para React](/learn/editor-setup#linting), verificará que cada valor reactivo esté correctamente especificado como dependencia. La lista de dependencias tienen que tener un número constante de elementos y que sean escritos en línea como `[dep1, dep2, dep3]`. React comparará cada dependencia con su valor previo usando el algoritmo de comparación [`Object.is`](https://developer.mozilla.org/es/docs/Web/JavaScript/Reference/Global_Objects/Object/is). Si no especificas ninguna dependencia, tu Effect se re-ejecutará después de cada re-renderizado del componente.

#### Retorno {/*returns*/}

`useInsertionEffect` devuelve `undefined`.

#### Advertencias {/*caveats*/}

> * Effect que sólo se ejecuta en el cliente. No se ejecutan durante el renderizado en el servidor.

Reviewer: ```suggestion
* Los Efectos sólo se ejecutan en el cliente. No se ejecutan durante el renderizado en el servidor.
```

* No puedes actualizar el estado dentro de `useInsertionEffect`.
* En el tiempo en que `useInsertionEffect` se ejecuta, las referencias aún no han sido acopladas y el DOM todavía no ha sido actualizado.

---

## Uso {/*usage*/}

> ### Inyectando estilos dinámicos desde librerías CSS-in-JS {/*injecting-dynamic-styles-from-css-in-js-libraries*/}

Reviewer: https://github.com/reactjs/es.reactjs.org/blob/main/TRANSLATION.md#el-gerundio

```suggestion
### Inyección de estilos dinámicos desde bibliotecas de CSS-in-JS {/*injecting-dynamic-styles-from-css-in-js-libraries*/}
```
